gcc/rust/checks/errors/rust-hir-pattern-analysis.cc

```diff
@@ -0,0 +1,1554 @@
+  HIR::TupleStructItems *elems = pattern->get_items ().get ();
+
+  std::vector<DeconstructedPat> fields;
+  switch (elems->get_item_type ())
+    {
+      case HIR::TupleStructItems::ItemType::MULTIPLE: {
+	HIR::TupleStructItemsNoRange *multiple
+	  = static_cast<HIR::TupleStructItemsNoRange *> (elems);
+
+	rust_assert (variant->get_fields ().size ()
+		     == multiple->get_patterns ().size ());
+	for (size_t i = 0; i < multiple->get_patterns ().size (); i++)
+	  {
+	    fields.push_back (
+	      lower_pattern (ctx, multiple->get_patterns ().at (i).get (),
+			     variant->get_fields ().at (i)->get_field_type ()));
+	  }
+	return DeconstructedPat (ctor, arity, fields, pattern->get_locus ());
+      }
+      break;
+      case HIR::TupleStructItems::ItemType::RANGED: {
+	// TODO: ranged tuple struct items
+	rust_unreachable ();
+      }
+      break;
+      default: {
+	rust_unreachable ();
+      }
+    }
+}
+
+static DeconstructedPat
+lower_struct_pattern (Resolver::TypeCheckContext *ctx,
+		      HIR::StructPattern *pattern, TyTy::VariantDef *variant,
+		      Constructor ctor)
+{
+  int arity = variant->get_fields ().size ();
+
+  // Initialize all field patterns to wildcard.
+  std::vector<DeconstructedPat> fields
+    = std::vector<DeconstructedPat> (arity, DeconstructedPat::make_wildcard (
+					      pattern->get_locus ()));
+
+  std::map<std::string, int> field_map;
+  for (int i = 0; i < arity; i++)
+    {
+      auto &f = variant->get_fields ().at (i);
+      field_map[f->get_name ()] = i;
+    }
+
+  // Fill in the fields with the present patterns.
+  HIR::StructPatternElements elems = pattern->get_struct_pattern_elems ();
+  for (auto &elem : elems.get_struct_pattern_fields ())
+    {
+      switch (elem->get_item_type ())
+	{
+	  case HIR::StructPatternField::ItemType::IDENT: {
+	    HIR::StructPatternFieldIdent *ident
+	      = static_cast<HIR::StructPatternFieldIdent *> (elem.get ());
+	    int field_idx
+	      = field_map.at (ident->get_identifier ().as_string ());
+	    fields.at (field_idx)
+	      = DeconstructedPat::make_wildcard (pattern->get_locus ());
+	  }
+	  break;
+	  case HIR::StructPatternField::ItemType::IDENT_PAT: {
+	    HIR::StructPatternFieldIdentPat *ident_pat
+	      = static_cast<HIR::StructPatternFieldIdentPat *> (elem.get ());
+	    int field_idx
+	      = field_map.at (ident_pat->get_identifier ().as_string ());
+	    fields.at (field_idx) = lower_pattern (
+	      ctx, ident_pat->get_pattern ().get (),
+	      variant->get_fields ().at (field_idx)->get_field_type ());
+	  }
+	  break;
+	  case HIR::StructPatternField::ItemType::TUPLE_PAT: {
+	    // TODO: tuple: pat
+	    rust_unreachable ();
+	  }
+	  break;
+	  default: {
+	    rust_unreachable ();
+	  }
+	}
+    }
+
+  return DeconstructedPat{ctor, arity, fields, pattern->get_locus ()};
+};
+
+static DeconstructedPat
+lower_pattern (Resolver::TypeCheckContext *ctx, HIR::Pattern *pattern,
+	       TyTy::BaseType *scrutinee_ty)
```

Comment:
All patterns in the match expr are lowered to [DeconstructedPat](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_pattern_analysis/pat/struct.DeconstructedPat.html).
We support struct and enum-variant patterns for now. Other unimplemented patterns are handled by converting them to wildcard.